layer/layer_store.go

```diff
@@ -106,11 +106,17 @@ func NewStoreFromGraphDriver(store MetadataStore, driver graphdriver.Driver) (St
 	return ls, nil
 }
 
-func (ls *layerStore) loadLayer(layer ChainID) (*roLayer, error) {
+func (ls *layerStore) loadLayer(layer ChainID) (l *roLayer, err error) {
 	cl, ok := ls.layerMap[layer]
 	if ok {
 		return cl, nil
 	}
+	defer func() {
+		// If failed to load the layer, remove the layer metadata.
+		if err != nil {
+			ls.store.Remove(layer)
+		}
+	}()
```

Comment:
Is removing the metadata really appropriate if the layer couldn't be loaded? I could see this being destructive if a temporary condition causes failures (for example, something is not mounted).

cc @dmcgowan